File: api/nas-scan.js
import fs from 'node:fs';
import path from 'node:path';

// Vite development adapter only. Set NAS_DEV_ROOT to override the local test
// root. Cloudflare Pages uses functions/api/nas-scan.js and the on-prem worker
// NAS_BASE_PATH, so this path is never used by the deployed site.
const DEFAULT_NAS_ROOT = process.env.NAS_DEV_ROOT || 'D:/Webmap/360 web mapping/Project_Test';

// --- shared helpers for stitched-folder counting + metadata CSV reads -------
function listImagesInStitchedFolder(fullFolder) {
  // Equirectangulars live either directly in the run folder (PTGui flat
  // output) or inside the `panoramas/` subfolder (tiled-cubic layout).
  try {
    const files = fs.readdirSync(fullFolder);
    const direct = files.filter((f) => /\.(jpe?g|png)$/i.test(f));
    if (direct.length > 0) return direct.sort();
    if (files.includes('panoramas')) {
      const panoDir = path.join(fullFolder, 'panoramas');
      if (fs.existsSync(panoDir)) {
        return fs.readdirSync(panoDir).filter((f) => /\.(jpe?g|png)$/i.test(f)).sort();
      }
    }
    return [];
  } catch {
    return [];
  }
}

function readMetadataRun(metadataSubgridDir, folderName) {
  // Returns { csvName, rows } for the metadata run folder when it exists.
  const metaFolder = path.join(metadataSubgridDir, folderName);
  if (!fs.existsSync(metaFolder)) return null;
  let csvName = '';
  let rows = 0;
  try {
    const metaFiles = fs.readdirSync(metaFolder);
    const csvFound = metaFiles.find((f) => f.endsWith('.csv') && !f.startsWith('003485-')) || metaFiles.find((f) => f.endsWith('.csv'));
    if (csvFound) {
      csvName = csvFound;
      const csvContent = fs.readFileSync(path.join(metaFolder, csvFound), 'utf-8');
      const lines = csvContent.trim().split(/\r?\n/).filter(Boolean);
      if (lines.length > 1) rows = lines.length - 1;
    }
  } catch {
    // ignore
  }
  return { csvName, rows };
}

function folderDateLabel(folderName) {
  // BP_20220630 / 20220904 → 2022-06-30 / 2022-09-04
  const dateMatch = folderName.match(/(\d{4})(\d{2})(\d{2})/);
  if (!dateMatch) return '';
  return `${dateMatch[1]}-${dateMatch[2]}-${dateMatch[3]}`;
}

export default async function handler(req, res) {
  try {
    const url = new URL(req.url, 'http://localhost');
    const action = url.searchParams.get('action') || 'subgrids';
    const basePath = url.searchParams.get('basePath') || DEFAULT_NAS_ROOT;
    const subgrid = (url.searchParams.get('subgrid') || 'N93E70').toUpperCase().trim();

    res.setHeader('Content-Type', 'application/json');

    // Action 1: List all subgrids in 03_Stitching and 00_Raw_data
    if (action === 'subgrids') {
      const stitchingDir = path.join(basePath, '03_Stitching', 'Project-OUT', 'Grid 1');
      const rawDir = path.join(basePath, '00_Raw_data', 'Grid 1');
      const metaDir = path.join(basePath, '01_Metadata', 'Grid 1');
      const detected = new Set();

      const scanDir = (target) => {
        if (fs.existsSync(target)) {
          try {
            fs.readdirSync(target, { withFileTypes: true })
              .filter((dirent) => dirent.isDirectory() && !dirent.name.startsWith('.'))
              .forEach((d) => detected.add(d.name.toUpperCase().trim()));
          } catch {
            // ignore
          }
        }
      };

      scanDir(stitchingDir);
      scanDir(rawDir);
      scanDir(metaDir);

      const detectedSubgrids = Array.from(detected).sort();

      const subgridList = detectedSubgrids.map((code) => {
        const inStitching = fs.existsSync(path.join(stitchingDir, code));
        return {
          code,
          existsInStitching: inStitching,
          label: code
        };
      });

      res.statusCode = 200;
      res.end(
        JSON.stringify({
          success: true,
          basePath,
          detectedSubgrids,
          subgrids: subgridList
        })
      );
      return;
    }

    // Action 2: Read exact survey folders inside 03_Stitching for selected subgrid
    if (action === 'survey-folders') {
      const subgridStitchingDir = path.join(basePath, '03_Stitching', 'Project-OUT', 'Grid 1', subgrid);
      const metadataSubgridDir = path.join(basePath, '01_Metadata', 'Grid 1', subgrid);

      if (!fs.existsSync(subgridStitchingDir)) {
        // Subgrid directory does not exist on disk yet
        res.statusCode = 200;
        res.end(
          JSON.stringify({
            success: true,
            subgrid,
            existsOnDisk: false,
            folders: []
          })
        );
        return;
      }

      const folderEntries = fs
        .readdirSync(subgridStitchingDir, { withFileTypes: true })
        .filter((d) => d.isDirectory() && !d.name.startsWith('.'))
        .map((d) => d.name);

      const folders = folderEntries.map((folderName) => {
        const fullFolder = path.join(subgridStitchingDir, folderName);
        let count = 0;
        let samplePano = '';
        let formatType = 'Equirectangular 360°';
        let formatDesc = 'Stitched single JPG + manifest.json';

        try {
          const files = fs.readdirSync(fullFolder);
          const images = files.filter((f) => /\.(jpe?g|png)$/i.test(f));
          count = images.length;

          if (images.length > 0) {
            samplePano = `${images[0]} .. ${images[images.length - 1]}`;
          } else {
            // Check for subfolders like panoramas / panorama-tiles
            const subdirs = files.filter((f) => {
              try {
                return fs.statSync(path.join(fullFolder, f)).isDirectory();
              } catch {
                return false;
              }
            });
            if (subdirs.includes('panorama-tiles') || subdirs.includes('panoramas')) {
              formatType = 'Tiled Cubic (Deep Zoom)';
              formatDesc = 'Multi-resolution tiles + backups';
              samplePano = 'panoramas/ & panorama-tiles/';
              // Real count only: count equirectangular images inside panoramas/
              // when present. Tile output reports 0 rather than an estimate.
              const panoDir = path.join(fullFolder, 'panoramas');
              if (fs.existsSync(panoDir)) {
                try {
                  count = fs.readdirSync(panoDir).filter((f) => /\.(jpe?g|png)$/i.test(f)).length;
                } catch {
                  count = 0;
                }
              } else {
                count = 0;
              }
            }
          }
        } catch {
          count = 0;
        }

        // Check for matching CSV in 01_Metadata
        let csvName = `${folderName}.csv`;
        let gpsCount = count || 0;
        const metaFolder = path.join(metadataSubgridDir, folderName);
        if (fs.existsSync(metaFolder)) {
          try {
            const metaFiles = fs.readdirSync(metaFolder);
            const csvFound = metaFiles.find((f) => f.endsWith('.csv') && !f.startsWith('003485-')) || metaFiles.find((f) => f.endsWith('.csv'));
            if (csvFound) {
              csvName = csvFound;
              const csvContent = fs.readFileSync(path.join(metaFolder, csvFound), 'utf-8');
              const lines = csvContent.trim().split(/\r?\n/).filter(Boolean);
              if (lines.length > 1) {
                gpsCount = lines.length - 1; // excluding header
              }
            }
          } catch {
            // ignore
          }
        }

        // Parse date from folder name if format YYYYMMDD
        let rawDate = '';
        let displayDate = folderName;
        const dateMatch = folderName.match(/(\d{4})(\d{2})(\d{2})/);
        if (dateMatch) {
          rawDate = `${dateMatch[1]}-${dateMatch[2]}-${dateMatch[3]}`;
          const dObj = new Date(`${dateMatch[1]}-${dateMatch[2]}-${dateMatch[3]}`);
          if (!isNaN(dObj.getTime())) {
            displayDate = dObj.toLocaleDateString('en-GB', { day: '2-digit', month: 'short', year: 'numeric' });
          }
        }

        return {
          id: folderName,
          name: folderName,
          displayDate,
          rawDate,
          panoramasCount: count,
          samplePano: samplePano,
          csvName,
          gpsCount,
          formatType,
          formatDesc,
          path: `/03_Stitching/Project-OUT/Grid 1/${subgrid}/${folderName}/`
        };
      });

      res.statusCode = 200;
      res.end(
        JSON.stringify({
          success: true,
          subgrid,
          existsOnDisk: true,
          folders
        })
      );
      return;
    }

    // Action 3: Read and parse actual survey CSV from 01_Metadata
    if (action === 'read-csv') {
      const folderName = url.searchParams.get('folder') || '';
      const customCsvName = url.searchParams.get('csv') || '';
      const metadataSubgridDir = path.join(basePath, '01_Metadata', 'Grid 1', subgrid);
      const possibleDirs = [
        path.join(metadataSubgridDir, folderName),
        metadataSubgridDir,
        path.join(basePath, '01_Metadata', subgrid)
      ];

      let csvPath = '';
      for (const d of possibleDirs) {
        if (fs.existsSync(d)) {
          if (customCsvName && fs.existsSync(path.join(d, customCsvName))) {
            csvPath = path.join(d, customCsvName);
            break;
          }
          const files = fs.readdirSync(d);
          const found = files.find((f) => f.endsWith('.csv') && !f.startsWith('003485-')) || files.find((f) => f.endsWith('.csv'));
          if (found) {
            csvPath = path.join(d, found);
            break;
          }
        }
      }

      if (!csvPath || !fs.existsSync(csvPath)) {
        res.statusCode = 404;
        res.end(JSON.stringify({ success: false, error: 'CSV file not found on disk' }));
        return;
      }

      const content = fs.readFileSync(csvPath, 'utf-8');
      const lines = content.trim().split(/\r?\n/).map((l) => l.trim()).filter(Boolean);
      if (lines.length < 2) {
        res.statusCode = 200;
        res.end(JSON.stringify({ success: true, records: [] }));
        return;
      }

      const header = lines[0].toLowerCase().split(/[,\t]/).map((h) => h.trim());
      const latIdx = header.findIndex((h) => h.includes('lat'));
      const lonIdx = header.findIndex((h) => h.includes('lon') || h.includes('lng'));
      const fileIdx = header.findIndex((h) => h.includes('file') || h.includes('name') || h.includes('img'));
      const headIdx = header.findIndex((h) => h.includes('head') || h.includes('yaw') || h.includes('azimuth'));
      const timeIdx = header.findIndex((h) => h.includes('time') || h.includes('date'));
      const distIdx = header.findIndex((h) => h.includes('distance'));

      const records = [];
      let rowTotal = 0;
      let skippedNoCoords = 0;
      for (let i = 1; i < lines.length; i++) {
        const parts = lines[i].split(/[,\t]/).map((p) => p.trim());
        if (parts.length < 2) continue;
        rowTotal += 1;
        const seqStr = String(i).padStart(4, '0');
        const lat = latIdx >= 0 ? parseFloat(parts[latIdx]) : NaN;
        const lon = lonIdx >= 0 ? parseFloat(parts[lonIdx]) : NaN;
        const heading = headIdx >= 0 ? parseFloat(parts[headIdx]) || 0 : 0;
        const timeVal = timeIdx >= 0 && parts[timeIdx] ? parts[timeIdx] : '';
        const distRaw = distIdx >= 0 && parts[distIdx] ? parseFloat(parts[distIdx]) : NaN;
        // The source filename comes only from the CSV itself. No filename is
        // invented here: rows without one stay empty and are reported as
        // unpaired instead of being silently fabricated.
        const srcName = fileIdx >= 0 && parts[fileIdx] ? parts[fileIdx] : '';
        // The CSV's own filename IS the canonical metadata name — the rename
        // target. The row-index synthetic name is only a fallback for CSVs
        // that carry no filename column (rig sequences can start at any
        // index, e.g. BP_20220630 begins at N93E70-0093.jpg).
        const targetName = srcName || `${subgrid}-${seqStr}.jpg`;
        // Metadata distance-to-previous (metres) — the survey's own
        // trajectory length source, preferred over recomputing from GPS.
        const distanceToPrevious = Number.isFinite(distRaw) ? distRaw : null;

        if (!isNaN(lat) && !isNaN(lon)) {
          records.push({
            index: i,
            sourceFilename: srcName,
            targetFilename: targetName,
            timestamp: timeVal,
            latitude: parseFloat(lat.toFixed(6)),
            longitude: parseFloat(lon.toFixed(6)),
            heading: parseFloat(heading.toFixed(1)),
            distanceToPrevious,
            isMatched: false
          });
        } else {
          skippedNoCoords += 1;
        }
      }

      res.statusCode = 200;
      res.end(JSON.stringify({ success: true, csvPath: path.basename(csvPath), rowTotal, skippedNoCoords, records }));
      return;
    }

    // Action 4: List the actual image files inside a stitched survey folder.
    // Auto-pairing uses this to verify metadata rows against real files.
    // PTGui output keeps the equirectangulars in a `panoramas/` subfolder,
    // so when direct images are absent we descend into it exactly like the
    // survey-folders counting does — otherwise the pairing stays unverified.
    if (action === 'folder-images') {
      const folderName = url.searchParams.get('folder') || '';
      if (subgrid.includes('..') || subgrid.includes('/') || subgrid.includes('\\') ||
          folderName.includes('..') || folderName.includes('/') || folderName.includes('\\')) {
        res.statusCode = 400;
        res.end(JSON.stringify({ success: false, error: 'Invalid path segment' }));
        return;
      }

      const target = path.join(basePath, '03_Stitching', 'Project-OUT', 'Grid 1', subgrid, folderName);
      if (!folderName || !fs.existsSync(target)) {
        res.statusCode = 200;
        res.end(JSON.stringify({ success: true, existsOnDisk: false, count: 0, images: [] }));
        return;
      }

      let files = [];
      try {
        files = fs.readdirSync(target);
      } catch {
        files = [];
      }

      let images = files.filter((f) => /\.(jpe?g|png)$/i.test(f)).sort();
      if (images.length === 0 && files.includes('panoramas')) {
        const panoDir = path.join(target, 'panoramas');
        try {
          images = fs
            .readdirSync(panoDir)
            .filter((f) => /\.(jpe?g|png)$/i.test(f))
            .sort();
        } catch {
          images = [];
        }
      }

      res.statusCode = 200;
      res.end(JSON.stringify({ success: true, existsOnDisk: true, count: images.length, images }));
      return;
    }

    // Action 5: List the FINAL panoramic images for one survey run
    // (05_Final/Project-OUT/Grid 1/<subgrid>/<folder>). The one-click bucket
    // upload only unlocks when this dataset actually exists on disk.
    if (action === 'final-images') {
      const folderName = url.searchParams.get('folder') || '';
      if (subgrid.includes('..') || subgrid.includes('/') || subgrid.includes('\\') ||
          folderName.includes('..') || folderName.includes('/') || folderName.includes('\\')) {
        res.statusCode = 400;
        res.end(JSON.stringify({ success: false, error: 'Invalid path segment' }));
        return;
      }

      const finalRunDir = path.join(basePath, '05_Final', 'Project-OUT', 'Grid 1', subgrid, folderName);
      if (!folderName || !fs.existsSync(finalRunDir)) {
        res.statusCode = 200;
        res.end(JSON.stringify({ success: true, existsOnDisk: false, count: 0, images: [], path: `/05_Final/Project-OUT/Grid 1/${subgrid}/${folderName || '{folder}'}/` }));
        return;
      }

      const images = listImagesInStitchedFolder(finalRunDir);
      res.statusCode = 200;
      res.end(JSON.stringify({
        success: true,
        existsOnDisk: true,
        count: images.length,
        images,
        path: `/05_Final/Project-OUT/Grid 1/${subgrid}/${folderName}/`
      }));
      return;
    }

    // Action 6: Daily processing registry — parent subgrids with child survey
    // runs (union of stitching / metadata / raw folders), each child carrying
    // its metadata row count + processed image count for the 4-PC pipeline.
    if (action === 'registry') {
      const gridsRoot = basePath;
      const stageRoots = {
        stitching: path.join(gridsRoot, '03_Stitching', 'Project-OUT', 'Grid 1'),
        metadata: path.join(gridsRoot, '01_Metadata', 'Grid 1'),
        raw: path.join(gridsRoot, '00_Raw_data', 'Grid 1')
      };

      const detected = new Set();
      Object.values(stageRoots).forEach((root) => {
        if (fs.existsSync(root)) {
          try {
            fs.readdirSync(root, { withFileTypes: true })
              .filter((dirent) => dirent.isDirectory() && !dirent.name.startsWith('.'))
              .forEach((d) => detected.add(d.name.toUpperCase().trim()));
          } catch {
            // ignore
          }
        }
      });

      const registry = Array.from(detected).sort().map((code) => {
        const stitchDir = path.join(stageRoots.stitching, code);
        const metaDir = path.join(stageRoots.metadata, code);
        const rawDir = path.join(stageRoots.raw, code);

        const runNames = new Set();
        [stitchDir, metaDir, rawDir].forEach((root) => {
          if (!fs.existsSync(root)) return;
          try {
            fs.readdirSync(root, { withFileTypes: true })
              .filter((d) => d.isDirectory() && !d.name.startsWith('.'))
              .forEach((d) => runNames.add(d.name));
          } catch {
            // ignore
          }
        });

        const children = Array.from(runNames).sort().map((name) => {
          const stitchRunDir = path.join(stitchDir, name);
          const stitched = fs.existsSync(stitchRunDir);
          const images = stitched ? listImagesInStitchedFolder(stitchRunDir).length : 0;
          const meta = readMetadataRun(metaDir, name);
          const stitchingPath = `/03_Stitching/Project-OUT/Grid 1/${code}/${name}/`;
          return {
            name,
            date: folderDateLabel(name),
            stitched,
            images,
            metadataRows: meta ? meta.rows : 0,
            csvName: meta ? meta.csvName : '',
            hasMetadata: !!meta,
            stitchingPath
          };
        });

        return {
          subgrid: code,
          existsInStitching: fs.existsSync(stitchDir),
          totals: {
            surveys: children.length,
            stitchedRuns: children.filter((c) => c.stitched).length,
            metadataTotal: children.reduce((acc, c) => acc + c.metadataRows, 0),
            imagesTotal: children.reduce((acc, c) => acc + c.images, 0)
          },
          children
        };
      });

      res.statusCode = 200;
      res.end(JSON.stringify({ success: true, registry }));
      return;
    }

    res.statusCode = 400;
    res.end(JSON.stringify({ error: `Unknown action: ${action}` }));
  } catch (err) {
    res.statusCode = 500;
    res.end(JSON.stringify({ error: err.message || String(err) }));
  }
}

File: worker/nas_scan.py
"""Survey folder, metadata, and image scans rooted at the configured NAS base."""
from __future__ import annotations

import csv
import os
import re
from datetime import datetime
from pathlib import Path
from typing import Any

IMAGE_EXTENSIONS = {".jpg", ".jpeg", ".png"}


def _safe_path(base: str, *parts: str) -> Path:
    root = Path(base).resolve()
    candidate = root.joinpath(*parts).resolve()
    if candidate != root and root not in candidate.parents:
        raise ValueError("Path escapes NAS_BASE_PATH.")
    return candidate


def _safe_segment(value: str, label: str) -> str:
    clean = (value or "").strip()
    if not clean or clean in {".", ".."} or "/" in clean or "\\" in clean:
        raise ValueError(f"Invalid {label} path segment.")
    return clean


def _image_names(folder: Path) -> list[str]:
    """List actual panorama images in a run folder or its panoramas child."""
    if not folder.is_dir():
        return []
    try:
        direct = sorted(p.name for p in folder.iterdir() if p.is_file() and p.suffix.lower() in IMAGE_EXTENSIONS)
        if direct:
            return direct
        pano_dir = folder / "panoramas"
        if pano_dir.is_dir():
            return sorted(p.name for p in pano_dir.iterdir() if p.is_file() and p.suffix.lower() in IMAGE_EXTENSIONS)
    except OSError:
        return []
    return []


def _metadata_run(meta_root: Path, run_name: str) -> tuple[str, int]:
    folder = meta_root / run_name
    if not folder.is_dir():
        return "", 0
    try:
        names = sorted(p.name for p in folder.iterdir() if p.is_file() and p.suffix.lower() == ".csv")
    except OSError:
        return "", 0
    csv_name = next((n for n in names if not n.startswith("003485-")), names[0] if names else "")
    if not csv_name:
        return "", 0
    try:
        with (folder / csv_name).open("r", encoding="utf-8-sig", newline="") as handle:
            return csv_name, sum(1 for _ in csv.DictReader(handle))
    except (OSError, csv.Error, UnicodeError):
        return csv_name, 0


def _date_labels(run_name: str) -> tuple[str, str]:
    match = re.search(r"(\d{4})(\d{2})(\d{2})", run_name)
    if not match:
        return "", run_name
    raw = f"{match.group(1)}-{match.group(2)}-{match.group(3)}"
    try:
        parsed = datetime.strptime(raw, "%Y-%m-%d")
        display = f"{parsed.day:02d} {parsed.strftime('%b %Y')}"
    except ValueError:
        display = run_name
    return raw, display


def _subgrid_roots(base: str, subgrid: str) -> dict[str, Path]:
    sg = _safe_segment(subgrid, "subgrid")
    return {
        "stitching": _safe_path(base, "03_Stitching", "Project-OUT", "Grid 1", sg),
        "metadata": _safe_path(base, "01_Metadata", "Grid 1", sg),
        "raw": _safe_path(base, "00_Raw_data", "Grid 1", sg),
    }


def scan_nas(
    base_path: str,
    action: str,
    subgrid: str = "",
    folder: str = "",
    csv_name: str = "",
) -> dict[str, Any]:
    """Implement the dashboard's `/api/nas-scan` action contract on the NAS."""
    base = str(Path(base_path).resolve())
    action = (action or "subgrids").strip().lower()

    if action == "subgrids":
        roots = [
            _safe_path(base, "03_Stitching", "Project-OUT", "Grid 1"),
            _safe_path(base, "00_Raw_data", "Grid 1"),
            _safe_path(base, "01_Metadata", "Grid 1"),
        ]
        found: dict[str, str] = {}
        for root in roots:
            if not root.is_dir():
                continue
            try:
                for entry in root.iterdir():
                    if entry.is_dir() and not entry.name.startswith("."):
                        found.setdefault(entry.name.upper().strip(), entry.name)
            except OSError:
                continue
        stitching_root = roots[0]
        items = [
            {
                "code": code,
                "existsInStitching": (stitching_root / actual).is_dir(),
                "label": code,
            }
            for code, actual in sorted(found.items())
        ]
        return {"success": True, "basePath": base, "detectedSubgrids": [x["code"] for x in items], "subgrids": items}

    if action == "survey-folders":
        sg = _safe_segment(subgrid, "subgrid")
        stitch_root = _safe_path(base, "03_Stitching", "Project-OUT", "Grid 1", sg)
        metadata_root = _safe_path(base, "01_Metadata", "Grid 1", sg)
        if not stitch_root.is_dir():
            return {"success": True, "subgrid": sg, "existsOnDisk": False, "folders": []}
        try:
            run_names = sorted(p.name for p in stitch_root.iterdir() if p.is_dir() and not p.name.startswith("."))
        except OSError:
            run_names = []
        folders = []
        for run_name in run_names:
            run_dir = stitch_root / run_name
            images = _image_names(run_dir)
            csv_name, metadata_rows = _metadata_run(metadata_root, run_name)
            raw_date, display_date = _date_labels(run_name)
            dirs = {p.name for p in run_dir.iterdir() if p.is_dir()} if run_dir.is_dir() else set()
            is_tiled = "panoramas" in dirs or "panorama-tiles" in dirs
            sample = f"{images[0]} .. {images[-1]}" if images else "panoramas/ & panorama-tiles/" if is_tiled else ""
            folders.append({
                "id": run_name,
                "name": run_name,
                "displayDate": display_date,
                "rawDate": raw_date,
                "panoramasCount": len(images),
                "samplePano": sample,
                "csvName": csv_name or f"{run_name}.csv",
                "gpsCount": metadata_rows if csv_name else len(images),
                "formatType": "Tiled Cubic (Deep Zoom)" if is_tiled else "Equirectangular 360°",
                "formatDesc": "Multi-resolution tiles + backups" if is_tiled else "Stitched single JPG + manifest.json",
                "path": f"/03_Stitching/Project-OUT/Grid 1/{sg}/{run_name}/",
            })
        return {"success": True, "subgrid": sg, "existsOnDisk": True, "folders": folders}

    if action == "read-csv":
        return read_survey_csv(base, subgrid, folder, csv_name)

    if action == "folder-images":
        sg = _safe_segment(subgrid, "subgrid")
        run_name = _safe_segment(folder, "folder")
        root = _safe_path(base, "03_Stitching", "Project-OUT", "Grid 1", sg, run_name)
        images = _image_names(root)
        return {"success": True, "existsOnDisk": root.is_dir(), "count": len(images), "images": images}

    if action == "final-images":
        sg = _safe_segment(subgrid, "subgrid")
        run_name = _safe_segment(folder, "folder")
        root = _safe_path(base, "05_Final", "Project-OUT", "Grid 1", sg, run_name)
        images = _image_names(root)
        return {
            "success": True,
            "existsOnDisk": root.is_dir(),
            "count": len(images),
            "images": images,
            "path": f"/05_Final/Project-OUT/Grid 1/{sg}/{run_name}/",
        }

    if action == "registry":
        roots = {
            "stitching": _safe_path(base, "03_Stitching", "Project-OUT", "Grid 1"),
            "metadata": _safe_path(base, "01_Metadata", "Grid 1"),
            "raw": _safe_path(base, "00_Raw_data", "Grid 1"),
        }
        subgrids: dict[str, str] = {}
        for root in roots.values():
            if not root.is_dir():
                continue
            try:
                for entry in root.iterdir():
                    if entry.is_dir() and not entry.name.startswith("."):
                        subgrids.setdefault(entry.name.upper().strip(), entry.name)
            except OSError:
                continue
        registry = []
        for code, actual in sorted(subgrids.items()):
            stitch_root = roots["stitching"] / actual
            meta_root = roots["metadata"] / actual
            raw_root = roots["raw"] / actual
            run_names: set[str] = set()
            for root in (stitch_root, meta_root, raw_root):
                if root.is_dir():
                    try:
                        run_names.update(p.name for p in root.iterdir() if p.is_dir() and not p.name.startswith("."))
                    except OSError:
                        pass
            children = []
            for run_name in sorted(run_names):
                stitch_run = stitch_root / run_name
                images = _image_names(stitch_run) if stitch_run.is_dir() else []
                csv_name, metadata_rows = _metadata_run(meta_root, run_name)
                raw_date, _ = _date_labels(run_name)
                children.append({
                    "name": run_name,
                    "date": raw_date,
                    "stitched": stitch_run.is_dir(),
                    "images": len(images),
                    "metadataRows": metadata_rows,
                    "csvName": csv_name,
                    "hasMetadata": bool(csv_name),
                    "stitchingPath": f"/03_Stitching/Project-OUT/Grid 1/{code}/{run_name}/",
                })
            registry.append({
                "subgrid": code,
                "existsInStitching": stitch_root.is_dir(),
                "totals": {
                    "surveys": len(children),
                    "stitchedRuns": sum(1 for c in children if c["stitched"]),
                    "metadataTotal": sum(c["metadataRows"] for c in children),
                    "imagesTotal": sum(c["images"] for c in children),
                },
                "children": children,
            })
        return {"success": True, "registry": registry}

    raise ValueError(f"Unknown NAS scan action: {action}")


def read_survey_csv(base_path: str, subgrid: str, folder: str, csv_name: str = "") -> dict[str, Any]:
    """Read and normalize a survey metadata CSV into dashboard pairing rows."""
    sg = _safe_segment(subgrid, "subgrid")
    run_name = _safe_segment(folder, "folder")
    meta_root = _safe_path(base_path, "01_Metadata", "Grid 1", sg)
    run_dir = meta_root / run_name
    if not run_dir.is_dir():
        raise FileNotFoundError("Metadata survey folder not found.")
    names = sorted(p.name for p in run_dir.iterdir() if p.is_file() and p.suffix.lower() == ".csv")
    chosen = csv_name if csv_name and csv_name in names else next((n for n in names if not n.startswith("003485-")), names[0] if names else "")
    if not chosen:
        raise FileNotFoundError("Survey CSV not found.")
    csv_path = run_dir / chosen
    with csv_path.open("r", encoding="utf-8-sig", newline="") as handle:
        sample = handle.read(2048)
        handle.seek(0)
        try:
            dialect = csv.Sniffer().sniff(sample, delimiters=",\t;")
        except csv.Error:
            dialect = csv.excel
        reader = csv.DictReader(handle, dialect=dialect)
        headers = {str(h or "").strip().lower(): h for h in (reader.fieldnames or [])}

        def find_header(*needles: str):
            for normalized, original in headers.items():
                if any(n in normalized for n in needles):
                    return original
            return None

        lat_key = find_header("lat")
        lon_key = find_header("lon", "lng")
        filename_key = find_header("file", "name", "img")
        heading_key = find_header("head", "yaw", "azimuth")
        time_key = find_header("time", "date")
        distance_key = find_header("distance")
        records = []
        row_total = 0
        skipped = 0
        for index, row in enumerate(reader, start=1):
            if not any(str(v or "").strip() for v in row.values()):
                continue
            row_total += 1
            try:
                lat = float(row.get(lat_key, "")) if lat_key else float("nan")
                lon = float(row.get(lon_key, "")) if lon_key else float("nan")
            except (TypeError, ValueError):
                lat, lon = float("nan"), float("nan")
            if lat != lat or lon != lon:
                skipped += 1
                continue
            src_name = str(row.get(filename_key, "") or "").strip() if filename_key else ""
            try:
                heading = float(row.get(heading_key, "") or 0) if heading_key else 0.0
            except (TypeError, ValueError):
                heading = 0.0
            try:
                distance = float(row.get(distance_key, "")) if distance_key and str(row.get(distance_key, "")).strip() else None
            except (TypeError, ValueError):
                distance = None
            timestamp = str(row.get(time_key, "") or "").strip() if time_key else ""
            seq = f"{index:04d}"
            target = src_name or f"{sg}-{seq}.jpg"
            records.append({
                "index": index,
                "sourceFilename": src_name,
                "targetFilename": target,
                "timestamp": timestamp,
                "latitude": round(lat, 6),
                "longitude": round(lon, 6),
                "heading": round(heading, 1),
                "distanceToPrevious": distance,
                "isMatched": False,
            })
    return {"success": True, "csvPath": chosen, "rowTotal": row_total, "skippedNoCoords": skipped, "records": records}
